### ISSUE TITLE
Phase 8: Implement sequential HandBelief TRAIN/VALIDATION experiments

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-arena#109]
Parent tracking: `lisbun/lisjong-project #18`
Phase refinement: `lisbun/lisjong-project #33`
Predecessor: `#107` / PR `#108`

# Summary

Implement the bounded Arena-side machinery for **Phase 8 — Sequential HandBelief experiments** exactly as refined in `lisbun/lisjong-project #33`, including its latest **Pre-implementation refinement lock** comment.

Phase 7 ended with formal `CONTINUE`:

```text
baseline TEST MAE  0.4947181847204438
snapshot TEST MAE  0.4847081506869042
Delta MAE          +0.010010034033539617
bootstrap 95% CI   [+0.009358140389365055, +0.01066210061949402]
positive games     10 / 10
```

Phase 8 does **not** reopen that TEST. It builds and compares two small sequential candidates using only TRAIN `100..139` and VALIDATION `140..149`.

Conceptually:

```text
locked Phase 5 TRAIN/VALIDATION
+ Phase 6 snapshot feature semantics
+ previous self-predicted HandBelief state
        ↓
S1 previous-belief recurrence
S2 previous-belief + learned-latent recurrence
        ↓
self-rollout VALIDATION comparison
+ rollout-depth diagnostics
+ physical-consistency gate
        ↓
select at most one Phase 9 candidate
        ↓
external reproducible Phase 8 artifact(s)
```

**Do not run learned inference on Phase 7 TEST seeds `150..159` while implementing or evaluating this Issue.**

---

# Preflight

Before implementation, refresh and treat as authoritative:

- `lisbun/lisjong-project #33` body and latest comments
- `lisbun/lisjong-project #18` latest Phase 7 checkpoint
- current `lisjong-arena` `main`
- repository root `AGENTS.md` and any path-specific instructions
- current `pyproject.toml` / ML dependency / CI workflow
- Phase 5 dataset model/builder/persistence/measurement seams
- Phase 6 feature/tensor/constraint/model/training/artifact seams
- Phase 7 TEST-seal/evaluation/result behavior

Refinement-time Arena main:

```text
a99aa74c9a6fa845555b1e0fc54f198a614ec2b0
```

Implementation must use current main at start rather than assuming this SHA remains current.

---

# Locked data identities

```text
raw corpus
779599517d787a9515c69a9bd8bd610a491520d24799b0532cd05e2d0136c79e

dataset
3167b61277a088f7c2b5c0e9e01aefb8af1b8c9c1609eb8de3f9a6bf7eff73e1
```

Formal partitions:

```text
TRAIN      seeds 100..139   18,890 TURN anchors
VALIDATION seeds 140..149    4,558 TURN anchors
TEST       seeds 150..159    4,726 TURN anchors  # forbidden in Phase 8
```

Dataset provenance:

```text
lisjong        6db1ddc0c6fae312801104008bf18660975f687d
lisjong-engine 8735e89e1aea000ab59368d0368d476787827741
lisjong-arena  e667890f0124670a6858fba13bc41767cdc80350
```

---

# Frozen snapshot reference

Reuse the Phase 6 frozen model only as the snapshot reference / compatibility authority.

```text
feature semantics
phase6-history-snapshot-v1

feature dimension
919

artifact logical identity
54513de7885b1d08f3df315ae255b89dd2dd0bad0eebfeaba554572a408d9089

weights SHA-256
8fabd711aad299d64dc665d0a989efe84876f8e956a36317907b5cf32eeea351

manifest SHA-256
99b32845f8fc875a01c13d5dd607c1558e9a1ec4dea4a903f0bf10be459e7d09
```

Locked snapshot VALIDATION reference:

```text
per-tile MAE  0.4863309527332531
per-hand L1   16.535252392930605
samples       4,558
```

Do not mutate the Phase 6 artifact.

---

# Critical boundary — TEST seal

Phase 7 has already exposed TEST `150..159`; it is no longer a fresh holdout and must not become Phase 8 development data.

Phase 8 orchestration must reject or exclude TEST **before sequential feature/state materialization or model-facing operations**.

No Phase 8:

- training
- validation
- candidate selection
- architecture choice
- threshold tuning
- rollout diagnostic
- unit/integration test
- CI job

may generate learned S1/S2 prediction for seeds `150..159`.

Synthetic TEST fixtures are allowed to prove the seal, but the real formal population is forbidden.

Phase 9 will receive a separately pre-registered fresh holdout only after Phase 8 candidate selection completes.

---

# Existing contracts to reuse

Do not create parallel identity or feature systems.

## Phase 5 example identity

`TurnExampleReference` already provides:

```text
GameIdentity
partition
round_index
checkpoint_index
anchor_index
hand_number
honba
round_revision
viewer_seat
```

## Phase 6 materialization

Reuse the partition-neutral/player-safe feature materialization seam and the exact Phase 6 feature/tensor semantics.

Phase 6 already materializes explicit:

```text
opponent_winds
row_marginals
column_marginals
expected-count target
```

Keep public feature generation independent of labels.

## Physical constraint

Reuse the exact Phase 6 `4 x 34` constrained allocation implementation.

Do not fork its math.

---

# Sequence construction

Add a dedicated Phase 8 sequence contract rather than changing the Phase 5 dataset schema.

Canonical sequence identity is exactly:

```text
(GameIdentity, round_index, viewer_seat)
```

Requirements:

- state never crosses game boundary
- state never crosses round boundary
- state never crosses viewer-seat boundary
- state never crosses partition boundary
- order within sequence is `checkpoint_index` ascending
- validate that this order is consistent with the original game-global `anchor_index`
- same input produces identical sequence identities/order

`hand_number`, `honba`, and `round_revision` remain readback/integrity checks, not a substitute sequence key.

---

# Previous-belief semantic axis

Never carry previous belief as anonymous row position.

Maintain recurrent opponent state as conceptually:

```text
Wind -> expected_count[34]
```

At each current anchor, remap that state into the current `opponent_winds` order explicitly.

Use the existing opponent-wind alignment checks where possible.

Previous expected-count inputs are scaled using the existing tile-count scale (`4.0`) or an equivalent fixed semantic scaling that is declared once and not fit from data.

---

# Candidate S1 — previous-belief recurrence

Use one fixed small model family.

Recommended exact baseline architecture:

```text
current Phase 6 feature: 919
previous predicted opponent counts: 102
concat: 1021
    -> Linear 128
    -> ReLU
    -> Linear 64
    -> ReLU
    -> Linear 136
    -> existing 4 x 34 physical constraint
```

If an objective implementation incompatibility requires changing this decomposition, stop before quantitative training and document/refine in #33; do not silently architecture-search.

Previous expected-count input at step `t` must be S1's own prediction from step `t-1`, keyed/remapped by Wind.

No exact previous target is selectable input.

---

# Candidate S2 — previous-belief + learned latent recurrence

Use one fixed small recurrent model family, not a model search.

Recommended exact architecture:

```text
step input:
    current Phase 6 feature 919
    + previous predicted opponent counts 102
    = 1021

GRU/GRUCell hidden size 128
    -> Linear 64
    -> ReLU
    -> Linear 136
    -> existing 4 x 34 physical constraint
```

Latent state resets to zero at every sequence boundary.

Previous HandBelief input remains the model's own prior prediction; the latent does not replace that explicit input.

No Transformer, attention model zoo, latent-size sweep, or HPO in this Issue.

---

# Initial recurrent state

At `depth=1` for each sequence:

- previous HandBelief input = existing conditional-uniform baseline prediction at that same public anchor
- S2 latent state = zeros

The baseline initialization is serving-safe/public-only and deterministic.

Do not initialize from exact hidden truth or exact previous label.

---

# Training semantics

Training loss remains the existing expected-count cell-weighted MSE on the constrained first three opponent rows.

Do not switch to sequence-macro loss merely because examples are grouped into sequences.

Forward recurrence must be self-predicted:

```text
prediction_t
    -> previous-belief input at t+1
```

Never:

```text
true_target_t
    -> previous-belief input at t+1
```

## Runtime defaults

Inherit Phase 6 defaults unless sequence batching requires a non-semantic adaptation:

```text
seed              0
dataloader seed   0
learning rate     1e-3
weight decay      0
max epochs        40
patience          6
workers           0
deterministic     true
torch threads     1
CPU-only
```

Use Adam as in Phase 6 unless current Phase 6 implementation proves otherwise at preflight.

## Checkpoint selection

Unlike the Phase 6 construction run, Phase 8's declared primary development metric is MAE.

Select checkpoint by **strictly lower pooled self-rollout VALIDATION per-tile MAE**.

- equal within `abs_tol=1e-12, rel_tol=0` does not replace the earlier checkpoint
- patience = 6
- TRAIN loss remains MSE

Do not select by TEST or by subgroup diagnostics.

---

# Pre-training sequence inventory / BPTT lock

Before the first S1/S2 quantitative training result is generated, produce a deterministic TRAIN/VALIDATION-only sequence inventory:

- sequence count per partition
- min/mean/median/max sequence length
- depth bucket counts
- confirmation that no TEST sequence was materialized

Choose BPTT policy solely from maximum sequence length:

```text
max sequence length <= 64
    -> full-sequence BPTT

max sequence length > 64
    -> truncation length 32
```

S1 and S2 must use the same policy.

If truncated:

- carry self-predicted belief across chunk boundaries
- carry S2 latent value across chunk boundaries
- detach gradient history only
- never reset to ground truth at a chunk boundary

Record the chosen policy in artifacts/results.

---

# VALIDATION evaluation

Run self-rollout from each sequence's deterministic initial state through every validation anchor.

Primary metric:

```text
pooled per-tile MAE
```

Reference effect:

```text
Delta MAE
= frozen Phase 6 snapshot VALIDATION MAE
- candidate VALIDATION MAE
```

Positive is better.

Also report:

- per-hand L1
- 10 per-game paired Delta MAE values vs frozen snapshot
- game-macro mean Delta MAE
- median per-game Delta MAE
- positive-game count
- physical consistency
- runtime / parameter count / throughput

## Locked rollout-depth diagnostic

Use exactly:

```text
depth 1
depth 2..4
depth 5..8
depth 9+
```

Depth is local to `(game, round, viewer)` sequence.

For every bucket report at minimum sample count and candidate MAE; where practical also snapshot MAE and delta.

Do not change bucket boundaries after results.

---

# Physical validity gate

A candidate cannot advance if VALIDATION has any:

```text
constraint non-convergence count != 0
max row/column residual > 1e-6
concealed-size inconsistency max > 1e-6
physical conservation violation sample rate != 0
```

Conservation total/mean excess remain mandatory report values under existing tolerance semantics.

No silent fallback/repair.

---

# Candidate advancement rule

Among technically valid S1/S2 candidates, lowest pooled VALIDATION MAE wins.

Advance to Phase 9 only if the winning candidate satisfies:

```text
Delta MAE > 0
positive validation-game count >= 6 / 10
physical validity passed
```

No Phase 8 materiality epsilon is imposed; Phase 9 is the fresh-holdout confirmatory model-family gate.

Tie rule:

```text
S1/S2 pooled VALIDATION MAE equal within abs_tol=1e-12, rel_tol=0
    -> prefer S1
```

If neither advances, record `no sequential candidate advances`; Phase 8 still completes and #18 receives a lightweight review.

---

# Artifact / result persistence

Persist external Git-unmanaged versioned artifacts/results for S1/S2.

At minimum bind:

- schema/version
- candidate semantics (`S1` / `S2`)
- model config
- parameter count
- weights SHA-256
- artifact logical identity
- raw/dataset identities
- exact TRAIN/VALIDATION population
- explicit `test_partition_evaluated=false`
- Phase 6 feature semantics ID
- sequence semantics/version
- previous-belief Wind-axis semantics
- initial-state semantics
- self-rollout semantics
- BPTT policy
- runtime/training config
- selected epoch
- TRAIN / VALIDATION metrics
- per-game diagnostics
- depth diagnostics
- physical metrics
- advancement result

Use state_dict-only strict load / `weights_only=True` / `map_location=cpu` style consistent with Phase 6 where applicable.

Refuse overwrite for formal candidate artifact destinations.

Do not mutate Phase 6 or Phase 7 artifacts.

Generated model/result files remain outside Git.

---

# CLI / execution boundary

Add a narrow explicit Phase 8 research entry point, e.g.:

```text
python -m lisjong_arena.phase8_sequential ...
```

Prefer explicit subcommands/boundaries such as:

```text
inventory
train-s1
train-s2
compare
```

or an equivalent fail-closed workflow.

Requirements:

- explicit raw/dataset paths
- explicit external output paths
- no TEST option
- no generic architecture/HPO flags
- no Phase 9 fresh-holdout generation/evaluation

Formal candidate training must not occur automatically on import, unit tests, CI, or PR creation.

---

# Tests

Use synthetic fixtures/temp artifacts. Do not run real learned S1/S2 inference on seeds `150..159`.

At minimum cover:

## Sequence identity / ordering

- exact `(GameIdentity, round_index, viewer_seat)` grouping
- checkpoint ordering
- anchor-order consistency
- reset across game/round/viewer
- partition isolation

## Previous-belief safety

- Wind-keyed previous state remaps correctly
- row-position swap cannot silently change opponent identity
- first-anchor initialization is public baseline only
- later anchors receive prior model prediction
- changing hidden labels cannot alter recurrent state before loss
- true previous target is never fed into selectable forward path

## TEST seal

- TRAIN/VALIDATION orchestration excludes/rejects real TEST partition before model-facing materialization
- inventory reports no TEST sequence
- formal CLI exposes no TEST evaluation path

## Determinism / rollout

- deterministic sequence inventory
- deterministic self-rollout for fixed weights/input
- chunked policy, if selected, preserves state values across chunks while detaching gradient only
- fixed depth buckets

## Metrics / selection

- pooled MAE semantics match existing expected-count measurement
- game deltas are paired against frozen snapshot on identical references
- physical invalidity prevents advancement
- S1/S2 winner/advancement/tie rules are exhaustive and deterministic

## Artifact

- strict identity/config/weights checks
- `test_partition_evaluated=false`
- overwrite refused
- Phase 6/7 artifact bytes unchanged

---

# CI

Extend the existing ML-focused CI pattern to include Phase 8 synthetic ML tests without performing formal training or TEST inference.

Normal Arena import/runtime must remain torch-free where the current architecture requires it.

Full repository CI remains merge authority.

---

# Known limitations

Current formal source population has no real coverage for:

```text
chi
pon
daiminkan
ankan
kakan
rinshan
non-discard response outcomes
```

Do not claim Phase 8 sequential quality for these populations.

Do not expand source/corpus to repair a disappointing VALIDATION result within this Issue.

---

# Non-goals

Do not implement:

- Phase 7 TEST reuse
- Phase 9 fresh holdout generation/evaluation
- AWS
- large-scale data generation
- HPO / model zoo
- Transformer
- event-level tokenizer redesign
- new HandBelief heads
- Mortal / human-log ingestion
- production estimator integration
- Decision consumer
- game-strength evaluation

---

# Acceptance criteria

- [ ] current main / #33 / AGENTS preflight refreshed
- [ ] deterministic sequence inventory implemented
- [ ] exact sequence key/order/reset semantics tested
- [ ] TEST sealed before any Phase 8 model-facing operation
- [ ] BPTT policy locked from sequence length only before candidate results
- [ ] S1 implemented with self-predicted previous HandBelief
- [ ] S2 implemented with self-predicted previous HandBelief + learned latent
- [ ] no exact previous label enters selectable recurrent input
- [ ] same Phase 6 physical constraint reused
- [ ] TRAIN-only weight updates
- [ ] VALIDATION-only checkpoint/candidate selection
- [ ] frozen snapshot VALIDATION compatibility reproduced
- [ ] pooled MAE/per-hand L1/per-game/depth/physical diagnostics recorded
- [ ] deterministic advancement/tie rule implemented
- [ ] S1/S2 external artifacts/results persisted with TEST=false
- [ ] runtime/parameter/throughput measurements recorded
- [ ] synthetic Phase 8 ML tests in CI
- [ ] full tests/lint/CI green
- [ ] no generated artifacts committed
- [ ] no Phase 9 work / Decision consumer / AWS scope creep

---

# Delivery

Implement → sequence-inventory/BPTT lock → candidate TRAIN/VALIDATION runs → self-review → Ready PR → GitHub Actions confirmation.

Do not merge automatically.

Report:

- sequence inventory and selected BPTT policy
- S1 metrics/artifact identity/runtime
- S2 metrics/artifact identity/runtime
- snapshot-vs-S1/S2 VALIDATION comparison
- advancement result
- PR URL / branch / head SHA
- CI status
- changed files
- any deviation from #33 and why

After merge, update #33/#18 and only then refine Phase 9 fresh-holdout protocol if a sequential candidate advances.
